#130341363 Add In-game chat feature

cyrielo opened the pull request:
#### What does this PR do?

This PR integrate a new game chat feature into the application
#### Description of Task to be completed?

When I am in a game, I can send messages to other users 
and view sent and received messages in a public chat window
#### How should this be manually tested?

Test suite have been written to test the functionality of the chat, execute the test and watch them pass
#### Any background context you want to provide?

The game chat is only functional when a game is being played
The story is not entirely completed, as there are missing functionality dependent on other service
#### What are the relevant pivotal tracker stories?

Relevant story:  #130341363 (Users should be able to chat with the group while they are in a game)
#### Screenshots (if appropriate)

None
#### Questions:

Commit:
chore(readme-update): update read me with new relevant info
Finishes #130373385

Commit:
Merge branch 'develop' of https://github.com/andela/project-kinsan into develop

Update files in develop.

Commit:
Merge branch 'develop' of https://github.com/andela/project-kinsan into develop

- Merge changes with local develop.

Commit:
Merge branch 'develop' of https://github.com/andela/project-kinsan into develop

- Update my develop branch for work on chat.

Commit:
feature(chat): in game chat

- Add branch for creating in game chat.

Start #130341363

Commit:
feature(chat):in game chat

Includes firebase reference and setup in foot.jade
add reference to css file in head.jade
add router to redirect to the game chat screen
add a game-chat controller
add a game chat service
add test for the game-chat

Start #130341363

Commit:
feature(chat): in game chat
- Add images for the chat widget
- Correct padding for elenents of the chat

#130341363

Commit:
feature(chat): in game chat
- Integrate front end with controller
- Implment <user> is typing feature
- Implement auto scroll to the newest chat
#130341363

Commit:
feature(chat): in game chat
-Remove border from message input in chat window
#130341363

Commit:
feature(chat): in game chat

-Update foot.jade to reference the firebase lib
-update bower dependency to reference firebase
-update karma.conf.js to be aware of firebase lib

Start #130341363

Commit:
feature(chat): in game

Refactor game-chat controller and integrate it to frontend
Refactor gamechat factory, add isTyping functionality
Complete writing tests for game-chat controller and game-chat factory

Start #130341363

Commit:
feature(chat): in game

-Resolve conflict in css file
-Update game.tpl.html to remove Doctype template
-Update game-chat.js controller to send message when enter key is pressed

Start #130341363

Commit:
feature(chat): in game chat

-Fix broken package.json
Start #130341363

Commit:
feature(chat): in game chat

Add fix for travis build failure

130341363

Commit:
feature(chat): in game chat

Fix travis build failure

130341363

Commit:
feature(chat): in game chat

Fix travis build failure

130341363

Commit:
Merge branch 'feature/130341363/in-game-chat' of https://github.com/andela/project-kinsan into feature/130341363/in-game-chat

Commit:
feature(chat): in game chat

Change for loop style

130341363

Commit:
feature(chat): in game chat

Fix travis build failure

130341363

Commit:
Merge branch 'feature/130341363/in-game-chat' of https://github.com/andela/project-kinsan into feature/130341363/in-game-chat
Try to fix travis build failure

Commit:
feature(chat): in game chat

Resolve HoundCI reported issues

Start 130341363

ruqoyyasadiq commented on app/views/includes/head.jade:
```diff
@@ -24,6 +24,7 @@ head
 
   link(rel='stylesheet', href='/lib/bootstrap/dist/css/bootstrap.css')
   link(rel='stylesheet', href='/css/landing.css')
+  link(rel='stylesheet', href='/css/chat.css')
```
Are we still writing `css`? The defined standard is to write `sass` which then compiles to `css`. This is not acceptable.

ruqoyyasadiq replied:
@andela-abisoye do fix this

ruqoyyasadiq replied:
@andela-abisoye can we have all these custom stylesheets (not 3rd party ones like bootstrap) compiled into one file called app.css? You can specify that in your gulp task to have all the numerous stylesheets (chat.scss, landing.scss and so on) bundled into one file after the transformation from scss to css. Please implement that and reach for further clarification if any.

ruqoyyasadiq commented on public/js/controllers/game-chat.js:
```diff
@@ -0,0 +1,78 @@
+'use strict';
+angular.module('mean.gameChat')
+  .controller('GameChatCtrl', function ($scope, GameChat) {
+    var gameId = 'rkhu3yj4',
```
when do we get to use the gameId value for each game?

Commit:
feature(chat): in-game-chat

- Convert css to sass/scss
- add folder to gulp task

#130341363

Commit:
Sort out conflict merge

Commit:
merge

Commit:
feature(chat): in-game-chat

- Make chat fixed with no retracting feature.
- Change color to blue

#130341363

Commit:
feature(chat): in-game-chat

- Make changes to pass hound test.

#130341363

Commit:
Hound Changes

Commit:
Fix Chat Text

Commit:
feature(chat): in game chat

Update karma.conf.js to load socket.io.js
Update app.js to add mean.system as dependency
Add socket.io.js library

130341363

Commit:
feature(chat): in game chat

Separate controller test and service test

130341363

Commit:
feature(chat): in game chat

Remove hardcoded values in game-chat controller
Add isTyping feature to controller
Add Jquery code to controller
Fix typo error in game-chat service
Remove jquery code from game.tpl.html

130341363

Commit:
feature(chat): in game chat

Mock the game service in the game-chat-ctrl.test.js
Remove the socket.io.js library

130341363

ruqoyyasadiq commented on public/css/chat.css:
```diff
@@ -0,0 +1,120 @@
+.chat-box {
```
if this file is being generated when gulp is run, then you should be ignoring it since it would get created when gulp tasks run no matter the environment.

ruqoyyasadiq replied:
@andela-abisoye take note 👆

ruqoyyasadiq commented on public/css/chat.scss:
```diff
@@ -0,0 +1,163 @@
+$grey-shadow-radius: 0 -1px 2px 0;
+$grey-shadow-75: rgba(65, 65, 65, 0.75);
+$box-radius: 5px 5px 0 0;
+$chat-bottom-padding: 0 5px 0 10px;
```
For future sake, we may want to have a variable.scss file to store the variables which could be reused across different stylesheets.

ruqoyyasadiq commented on public/css/chat.scss:
```diff
@@ -0,0 +1,163 @@
+
+$url-0: url("../img/chatnone1.png");
+$url-1: url("../img/chatnone.png");
+
```
same with these variables too.

ruqoyyasadiq commented:
@cyrielo and @andela-abisoye fix these comments and let's merge in!

Commit:
feature(game): in-game-chat
- Remove css files from repo and from being tracked.
- Create seperate scss file for variables called variables.scss

#130341363

ruqoyyasadiq commented on public/css/chat.scss:
```diff
@@ -0,0 +1,150 @@
+@import "variable.scss";
```
remove the `.scss` in the import statement.

ruqoyyasadiq replied:
@andela-abisoye 👆

Commit:
feature(game): in-game-chat
- Make Hound Changes.

#130341363

ruqoyyasadiq reviewed (approved):
Approved
